server/etcdserver/server.go

```diff
@@ -2170,6 +2180,22 @@ func (s *EtcdServer) snapshot(snapi uint64, confState raftpb.ConfState) {
 		"saved snapshot",
 		zap.Uint64("snapshot-index", snap.Metadata.Index),
 	)
+}
+
+func (s *EtcdServer) maybeCompactRaftLog(ep *etcdProgress) {
+	// Retain all log entries up to the latest snapshot index to ensure any member can recover from that snapshot.
+	// Beyond the snapshot index, preserve the most recent s.Cfg.SnapshotCatchUpEntries entries in memory.
+	// This allows slow followers to catch up by synchronizing entries instead of requiring a full snapshot transfer.
+	if ep.snapi <= s.Cfg.SnapshotCatchUpEntries {
+		return
+	}
+
+	compacti := ep.snapi - s.Cfg.SnapshotCatchUpEntries
```

Comment:
```suggestion
	compacti := ep.appliedi - s.Cfg.SnapshotCatchUpEntries
```

Reply:
What's the purpose of the change?
If you use `ep.snapi`, then the behaviour is exactly the same as existing behaviour, because `ep.snapi` only gets updated each time after creating the (v2) snapshot.

Reply:
Please see https://github.com/etcd-io/etcd/issues/17098#issuecomment-2374327464. Can we get that confirmed firstly?

Reply:
This change doesn't change when compaction is run or how many times it's executed. I was aware of https://github.com/etcd-io/raft/blob/5d6eb55c4e6929e461997c9113aba99a5148e921/storage.go#L266-L269 code, that's why I was proposing compacting only ever X applies.